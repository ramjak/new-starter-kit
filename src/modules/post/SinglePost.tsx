import React, { useCallback } from 'react';
import styles from './SinglePost.module.scss';
import { Button } from '@material-ui/core';
import { Delete, Edit } from '@material-ui/icons';
import IPost from '../../domains/post';
import ROUTES, { Link, useNavigateTo } from '../../routes';

interface ISinglePost {
  item: IPost;
  destroy(id: string): Promise<any>;
}

export function SinglePost({ item, destroy }: ISinglePost) {
  const navigateTo = useNavigateTo();

  const navigateToEditForm = useCallback(
    () => navigateTo(ROUTES.editSinglePost, item.id),
    [navigateTo, item]
  );
  const destroyThis = useCallback(() => destroy(item.id.toString()), [
    destroy,
    item,
  ]);

  return (
    <article className={styles.post}>
      <h2>
        <Link route={ROUTES.viewSinglePost} params={[item.id]}>
          {item.title}
        </Link>
      </h2>
      <section title="action buttons" className={styles.actions}>
        <Button variant="outlined" color="default" onClick={navigateToEditForm}>
          <Edit />
        </Button>
        &nbsp;
        <Button variant="contained" color="secondary" onClick={destroyThis}>
          <Delete />
        </Button>
      </section>
      <p>{item.body}</p>
    </article>
  );
}
