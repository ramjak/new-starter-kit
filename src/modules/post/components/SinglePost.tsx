import React, { useCallback } from "react";
import { Button } from "@material-ui/core";
import { Delete, Edit } from "@material-ui/icons";
import styles from "./SinglePost.module.scss";
import IPost from "../../../domains/post";

interface ISinglePost {
  item: IPost;
  view(id: string): unknown;
  edit(id: string): unknown;
  destroy(id: string): Promise<unknown>;
}

export default function SinglePost({ item, destroy, edit, view }: ISinglePost) {
  const viewThis = useCallback(() => view(item.id.toString()), [view, item.id]);
  const editThis = useCallback(() => edit(item.id.toString()), [edit, item.id]);
  const destroyThis = useCallback(() => destroy(item.id.toString()), [
    destroy,
    item,
  ]);

  return (
    <article className={styles.post}>
      <Button
        className={styles.title}
        color="primary"
        onClick={viewThis}
        data-testid="view-btn"
        component="h2"
      >
        {item.title}
      </Button>
      <section title="action buttons" className={styles.actions}>
        <Button
          variant="outlined"
          color="default"
          onClick={editThis}
          data-testid="edit-btn"
        >
          <Edit />
        </Button>
        &nbsp;
        <Button
          variant="contained"
          color="secondary"
          onClick={destroyThis}
          data-testid="destroy-btn"
        >
          <Delete />
        </Button>
      </section>
      <p data-testid="article-body">{item.body}</p>
    </article>
  );
}
