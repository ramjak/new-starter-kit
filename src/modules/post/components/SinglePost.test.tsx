/* eslint-disable react/jsx-no-bind */
import "reflect-metadata";
import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render } from "@testing-library/react";
import SinglePost from "./SinglePost";

it("render SinglePost", () => {
  const result = renderer
    .create(
      <SinglePost
        item={{
          userId: 1,
          title: "title",
          id: 2,
          body:
            "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
        }}
        view={() => undefined}
        edit={() => undefined}
        destroy={() => Promise.resolve()}
      />
    )
    .toJSON();

  expect(result).toMatchSnapshot();
});

it("invoke functions on buttons click", () => {
  const viewFn = jest.fn();
  const editFn = jest.fn();
  const destroyFn = jest.fn();

  const { getByTestId } = render(
    <SinglePost
      item={{
        userId: 1,
        title: "title",
        id: 2,
        body:
          "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
      }}
      view={viewFn}
      edit={editFn}
      destroy={destroyFn}
    />
  );

  fireEvent.click(getByTestId("view-btn"));
  expect(viewFn).toBeCalledWith("2");

  fireEvent.click(getByTestId("edit-btn"));
  expect(editFn).toBeCalledWith("2");

  fireEvent.click(getByTestId("destroy-btn"));
  expect(destroyFn).toBeCalledWith("2");
});
