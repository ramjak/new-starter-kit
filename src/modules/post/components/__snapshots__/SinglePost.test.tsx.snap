// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render SinglePost 1`] = `
<article
  className="post"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text title MuiButton-textPrimary"
    data-testid="view-btn"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      title
    </span>
  </button>
  <section
    className="actions"
    title="action buttons"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined"
      data-testid="edit-btn"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
      </span>
    </button>
    Â 
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
      data-testid="destroy-btn"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </span>
    </button>
  </section>
  <p>
    quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto
  </p>
</article>
`;
